source/non_matching/mesh_classifier.cc

```diff
@@ -0,0 +1,541 @@
+      {}
+
+
+
+      template <int dim>
+      void
+      AnalyticLevelSetDescription<dim>::get_local_level_set_values(
+        const typename Triangulation<dim>::active_cell_iterator &cell,
+        Vector<double> &local_levelset_values)
+      {
+        fe_values.reinit(cell);
+        const std::vector<Point<dim>> &points =
+          fe_values.get_quadrature_points();
+
+        for (unsigned int i = 0; i < points.size(); i++)
+          local_levelset_values[i] = level_set->value(points[i]);
+      }
+
+
+
+      template <int dim>
+      void
+      AnalyticLevelSetDescription<dim>::get_local_level_set_values(
+        const typename Triangulation<dim>::active_cell_iterator &cell,
+        const unsigned int                                       face_index,
+        Vector<double> &local_levelset_values)
+      {
+        fe_face_values.reinit(cell, face_index);
+        const std::vector<Point<dim>> &points =
+          fe_face_values.get_quadrature_points();
+
+        for (unsigned int i = 0; i < points.size(); i++)
+          local_levelset_values[i] = level_set->value(points[i]);
+      }
+
+
+
+      template <int dim>
+      const hp::FECollection<dim> &
+      AnalyticLevelSetDescription<dim>::get_fe_collection() const
+      {
+        return fe_collection;
+      }
+
+
+
+      template <int dim>
+      unsigned int
+      AnalyticLevelSetDescription<dim>::active_fe_index(
+        const typename Triangulation<dim>::active_cell_iterator &) const
+      {
+        return 0;
+      }
+    } // namespace MeshClassifierImplementation
+  }   // namespace internal
+
+
+
+  using namespace internal::MeshClassifierImplementation;
+
+  template <int dim>
+  template <class VECTOR>
+  MeshClassifier<dim>::MeshClassifier(const DoFHandler<dim> &dof_handler,
+                                      const VECTOR &         level_set)
+    : triangulation(&dof_handler.get_triangulation())
+    , level_set_description(
+        std::make_unique<DiscreteLevelSetDescription<dim, VECTOR>>(dof_handler,
+                                                                   level_set))
+  {}
+
+
+
+  template <int dim>
+  MeshClassifier<dim>::MeshClassifier(const Triangulation<dim> &triangulation,
+                                      const Function<dim> &     level_set,
+                                      const FiniteElement<dim> &element)
+    : triangulation(&triangulation)
+    , level_set_description(
+        std::make_unique<AnalyticLevelSetDescription<dim>>(level_set, element))
+  {
+    // The level set function must be scalar.
+    AssertDimension(level_set.n_components, 1);
+    AssertDimension(element.n_components(), 1);
+  }
+
+
+
+  template <int dim>
+  void
+  MeshClassifier<dim>::reclassify()
+  {
+    initialize();
+
+    // Loop over all cells and determine category of all locally owned
+    // cells and faces.
+    for (const auto &cell : triangulation->active_cell_iterators())
+      {
+        if (cell->is_locally_owned())
```

Comment:
I would love if this could also be done for ghost cells,
```suggestion
        if (!cell->is_artificial())
```
I tried with this locally (as I want to have DG solutions also in the neighbor) and it works fine, as long as the vector has enough ghost entries updated.